### ISSUE TITLE
External provisioning problems

### ISSUE THREAD

**Issue by cofyc** [kubernetes/kubernetes#72031]
I did more investigations in issue https://github.com/kubernetes/kubernetes/issues/71928, I think there are two problems:

- scheduler has no way to notify external provisioner to retry provisioning on reschedule (e.g. invalid storage class is fixed)
- external provisioner has no way to notify scheduler to reschedule PVC when it encounters unrecoverable situations (e.g. selected-node is wrong)

Here are two scenarios:

#### Scenario 1) selected node is right, but need to notify external provisioner to provision because non-PVC objects (e.g. storage class) changed

This is case in this issue.

Possible ways to do:

1.1) update [external provisioner lib](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner) to retry claim on storage class events.

This requires to update external provisioner.

1.2) schedule PVC always when necessary, e.g. storage class or PVC is updated, no matter it's already being selected a node or not. But we must skip nodes except current `selected-node`, because it's not safe to change selected node by scheduler. A volume may be in creating on old node, we must relies on external provisioner to reject node for safety. Using this method, we need find a way to (e.g. add another annotation) to trigger PVC update event without changing current `selected-node`. 

This requires to update external provisioner to provision again immediately on PVC update event.

1.3) remove `selected-node` on bind timeout

We assume it's safe to remove `selected-node` after bind timeout. It may take some time (depends bind timeout), but will notify external provisioner to provision again eventually, even if external provisioner exceeded retry limits. 

This requires to update external provisioner to provision again immediately on PVC update event.

#### Scenario 2) selected node is wrong, we need notify scheduler to find a feasible node

Possible ways to do:

2.1) wait provisioner to reject PVC from given node by removing `selected-node` annotation, then scheduler can reset it to notify provisioner to retry again. 

This works like pod scheduling, when a Pod is assigned to a node, but kubelet finds it is not correct, it will be rejected, e.g. node selector not matched now.

We need to distinguish two kinds of failures in external provisioner:

- unrecoverable failures, e.g. capacity/volume count hard limits on this node
  - scheduler should retry on these failures, and it will and must select a new node for given PVC (otherwise, it may enter into infinite loop)
- recoverable failures, e.g. network error, provisioner bugs etc
  - provisioners should retry on these failures and wait network to recover or bugs to be fixed

A more brutal way is to remove `selected-node` annotation on all failures, like internal provisioner. Anyhow, it requires external provisioner to cooperate with scheduler.

2.2) Another way is to schedule PVC always when necessary. And assign PVC to new node to notify provisioner to provision again. But it's not safe to change selected node by scheduler, see method 1.2). And current external provisioner will not retry provision immediately on `selected-node` field changed. We need to update external provisioner too.

2.3) Same as 1.3)

### My suggestion 

For now, we can do 1.1), this method has no drawbacks IMO.

/kind feature
/sig storage

**Comment by cofyc** [comment]
cc @msau42  Does current or planning design cover these problems?

**Comment by msau42** [comment]
The current design is supposed to use the selectedNode annotation as a way for scheduler and PV controller (internal + external) to communicate.  Provisioner is supposed to remove the selectedNode annotation if provisioning fails for any reason.

**Comment by cofyc** [comment]
Then, I think we should do 2.1). We can support removing `selectedNode` annotation in  [external provisioner lib](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner) first.

Current, it has too kinds of errors:

- [IgnoredError](https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/master/controller/volume.go#L65), skip current action
- Normal errors, framework retry with exponential backOff by default (this is recoverable error as I think)

I think we can add another kind of error (UnrecoverableError), when this error is returned,  framework stops retrying provision and removes `selectedNode` annotation.  What do you think?

**Comment by msau42** [comment]
I think we can keep it simple for now and just remove the annotation for all errors (except ignorederror).  That way it behaves as close to internal provisioner as possible.  Provisioning based by capacity and count limits are features we're in the process of adding so will be handled better in the future.

**Comment by cofyc** [comment]
I agree with this simpler method, and am going to add this support in https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner.

**Comment by pohly** [comment]
/unassign

I've put https://github.com/kubernetes/enhancements/pull/1703 up for adoption by someone who wants to carry it forward, if that is the way we want to go.
